### ISSUE TITLE
Bug: [prefer-optional-chain] Rule is triggered even when the type of the variable would change

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#5269]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3mpBbEAyhjAJYoDmAXNsWefAD7wCuKoAZmSMANwCwAKEHJ0WXAE8AQkiQQQUFAEZ4AXmx5CtCvACEKtaw5dg8AGSn1+IqQoA6WrgAUASlsg2aAOokMAC0cAROwyAc78QgIimPAS0rLyKABMqpaaNuQA-PakTq7uwF4+-kEhYYJAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uigB3UdCaRwYAL4gNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
The linked playground has the following code:


const someString: string | undefined;

const myBoolean1 = someString !== undefined && someString.trim().endsWith("foo");

const myBoolean2 = someString?.trim().endsWith("foo");


When declaring `myBoolean1`, the `prefer-optional-chain` is triggered, which forces me to write the code instead as it is in `myBoolean2`. However, `myBoolean1` and `myBoolean2` are not the safe. The type of the former is `boolean`, and the type of the latter is `boolean | undefined`.

This is problematic for many reasons. For example, when combined with the `strict-boolean-expressions` rule, any subsequent reference to `myBoolean2` would have to be changed from:

```ts
if (myBoolean2) {
```

To:

```ts
if (myBoolean2 === true) {
```

That can't possibly be best practice, right? It feels like the rule itself must be bugged.
```


### ESLint Config

```javascript
.
```

### Expected Result

.

### Actual Result

.

### Versions

.

**Comment by bradzacher** [comment]
The type change is a known limitation as the rule does not use type information.

Duplicate of #4490, #2963

---

> That can't possibly be best practice, right?

When using `strict-boolean-expressions`, yes it is best practice to clearly write your code to differentiate the truthy/falsy states.

**Comment by Zamiell** [comment]
> The type change is a known limitation as the rule does not use type information.

Ah, too bad. I'll throw in my 2 cents that the rule should not be turned on by default in the strict ruleset until it has type information. (Since situations like the OP will lead to writing unsafe code.)

> When using strict-boolean-expressions, yes it is best practice to clearly write your code to differentiate the truthy/falsy states.

To clarify, my problem isn't with `strict-boolean-expressions`. It's with reworking code to satisfy the linter that has the side effect of making the type less precise, which triggers downstream rules including `strict-boolean-expressions`.

It's worth noting here that we can use the nullish coalescing operator to fix the problem in the OP:

```ts
const myBoolean2 = someString?.trim().endsWith("foo") ?? false;
```

This is certainly less characters than the original of:

```
const myBoolean1 = someString !== undefined && someString.trim().endsWith("foo");
```

But is it really best practice? In the latter example, the TypeScript compiler guides me to write correct code, and in the former example, `false` is provided as a "magic boolean" that has no type safety. Nothing prevents me from mistyping `false` as `true` here, so this doesn't really strike me as safer or better code.

**Comment by bradzacher** [comment]
It's not about making it safer - it's about standardisation.

The rule is dumb - it just looks for cases that look like they could be converted to optional chaining. There are a few different forms but the general case is that it looks for `a && a.b`.

The rule has to be dumb because it is impossible to predict how you are intending to use the value. There's no way to know if making the change has downstream effects either.
This is why the rule has no fixer - because it would be unsafe to blindly fix the report.

---

The provided configs serve only as a guide to help people bootstrap their codebase. If you feel a rule doesn't fit your coding style, feel free to disable it in your local config.

**Comment by bradzacher** [comment]
In your specific example you are enforcing some base case and then accessing a member that results in a `boolean`. So in this specific case `someString !== undefined && someString.trim().endsWith("foo")` is actually clear, correct, concise code.

But consider a more general example like `someString !== undefined && someString.length`. Now you have an expression which returns `boolean | number` - which is really weird, right? How do you get out of that state in a typesafe way? Well you'd need a `typeof` check - UGLY!
But if you instead used `someString?.length` - now you have `number | undefined`! Now the code is significantly easier to work with!

This is what I mean when I say the rule is dumb and has to be dumb!

---

As for your point about which code is better - it's entirely subjective.

There are people that believe `someString?.endsWith('foo') ?? false` is the most clear and correct code.
There are some that believe `someString?.endsWith('foo') === true` is the most clear and correct code.
There are some that believe `someString !== undefined && someString.endsWith('foo')` is the most clear and correct code.

Which is why [`strict-boolean-expressions` has a wide variety of suggestion fixers](https://typescript-eslint.io/rules/strict-boolean-expressions/#fixes-and-suggestions).

---

Relying solely on TS is one school of thought - though it doesn't always lead to the cleanest/clearest code. So whilst your code typechecks fine - it might be hard to work with, or different people might go in and tweak code to their own style as they change it (messing with the blame history and adding fud to PRs).

Many lint rules are there to bridge that gap by enforcing codestyle and consistency (we call them stylistic rules). They're there to make everyone code in the same way. This reduces the cognitive overhead of writing code (I write code and if there are no lints it's stylistically correct), it reduces the chance of arguments about code style (yes - people are very passionate and can have heated arguments about code style), and it reduces reviewer burden (did this diff pass the linter? Yes - then the code style is fine and I can focus on logic).
